monetdbe/_cffi/internal.py

```diff
@@ -259,31 +256,49 @@ def append(self, table: str, data: Mapping[str, np.ndarray], schema: str = 'sys'
             raise exceptions.ProgrammingError(error)
 
         work_columns = ffi.new(f'monetdbe_column * [{n_columns}]')
+        work_objs = []
         # cffi_objects assists to keep all in-memory native data structure alive during the execution of this call
-        cffi_objects = []
+        cffi_objects = list()
         for column_num, (column_name, existing_type) in enumerate(existing_columns):
             column_values = data[column_name]
             work_column = ffi.new('monetdbe_column *')
             type_info = numpy_monetdb_map(column_values.dtype)
 
             # try to convert the values if types don't match
             if type_info.c_type != existing_type:
-                precision_warning(type_info.c_type, existing_type)
-                to_numpy_type = monet_c_type_map[existing_type].numpy_type
-                try:
-                    column_values = column_values.astype(to_numpy_type)
-                    type_info = numpy_monetdb_map(column_values.dtype)
-                except Exception as e:
-                    existing_type_string = monet_c_type_map[existing_type].c_string_type
-                    error = f"Can't convert '{type_info.c_string_type}' " \
-                            f"to type '{existing_type_string}' for column '{column_name}': {e} "
-                    raise ValueError(error)
+                if type_info.c_type == lib.monetdbe_timestamp and existing_type == lib.monetdbe_date and np.issubdtype(column_values.dtype, np.datetime64):
+                    """
+                    We are going to cast to a monetdbe_date and
+                    consider monetdbe_timestamp as a 'base type' to signal this.
+                    """
+                    type_info = timestamp_to_date()
+                else:
+                    precision_warning(type_info.c_type, existing_type)
+                    to_numpy_type = monet_c_type_map[existing_type].numpy_type
+                    try:
+                        column_values = column_values.astype(to_numpy_type)
+                        type_info = numpy_monetdb_map(column_values.dtype)
+                    except Exception as e:
+                        existing_type_string = monet_c_type_map[existing_type].c_string_type
+                        error = f"Can't convert '{type_info.c_string_type}' " \
+                                f"to type '{existing_type_string}' for column '{column_name}': {e} "
+                        raise ValueError(error)
 
             work_column.type = type_info.c_type
             work_column.count = column_values.shape[0]
             work_column.name = ffi.new('char[]', column_name.encode())
-            if type_info.numpy_type.kind == 'U':
+            if type_info.numpy_type.kind == 'M':
+                t = ffi.new('monetdbe_data_timestamp[]', work_column.count)
```

Comment:
just a comment, no action required, but this function is getting a bit messy, we should split it up in the future.